File: newpca.m
% New work on PCA. 2017-11-07 by Weihua Wang
% Read and plot pca results from R. 
% SO2, NO2, HCHO, IR340/675, I500, PC1, PC2, PC3, PC4, PC5 and Time should be assigned.

close all,clear all,clc;
row=1,col=0; % 1st row with header corresponds to an index of 0. column 0 means the 1st column. 
dt = csvread('~/pcadat1.csv',row,col);  % ~/  means read from home directory
dt(1,:) % check if correctly imported
HCHO = dt(:,1); % 1st column refers to HCHO concentration
NO2 = dt(:,2);
SO2 = dt(:,3);
IR340_675 = dt(:,4);
I500 = dt(:,5);
PC1 = dt(:,6);
PC2 = dt(:,7);
PC3 = dt(:,8);
PC4 = dt(:,9);
PC5 = dt(:,10);
Time = dt(:,11);

% Rescaled and standardized plots
plot(Time, SO2/50, '-b','linewidth',2); 
xlabel('Time'); ylabel('');
title('Rescaled standardized SO2/NO2/HCHO, I500, IR340/675 and PC1');
xlim([Time(1), Time(end)]); ylim([-0.4,0.6]);
hold on;
plot(Time, NO2/30, '-g','linewidth',2);
plot(Time, HCHO/50, '-k','linewidth',2);
plot(Time, IR340_675/40, '-m','linewidth',2);
plot(Time, I500/6, '-c','linewidth',2);
plot(Time, PC1/20, '-r','linewidth',2);
legend('SO2conc / 50', 'NO2conc / 30', 'HCHOconc / 50',...
'IR_{340/675} / 40', 'I500 / 6', 'PC1 / 20');

%print Newpca1.pdf;
set(gcf, 'papersize', [17, 9.6], 'paperposition', [0 0 17 9.6]);
print Newpca1.pdf;

%

plot(Time, SO2/50, '-b','linewidth',2); 
xlabel('Time'); ylabel('');
title('Rescaled standardized SO2/NO2/HCHO, I500, IR340/675 and PC2');
xlim([Time(1), Time(end)]); ylim([-0.4,0.6]);
hold on;
plot(Time, NO2/30, '-g','linewidth',2);
plot(Time, HCHO/50, '-k','linewidth',2);
plot(Time, IR340_675/40, '-m','linewidth',2);
plot(Time, I500/6, '-c','linewidth',2);
plot(Time, PC2/30, '-r','linewidth',2);
legend('SO2conc / 50', 'NO2conc / 30', 'HCHOconc / 50',...
'IR_{340/675} / 40', 'I500 / 6', 'PC2 / 30');

set(gcf, 'papersize', [17, 9.6], 'paperposition', [0 0 17 9.6]);
print Newpca2.pdf;

%

plot(Time, SO2/35, '-b','linewidth',2); 
xlabel('Time'); ylabel('');
title('Rescaled standardized SO2/NO2/HCHO, I500, IR340/675 and PC3');
xlim([Time(1), Time(end)]); ylim([-0.4,0.8]);
hold on;
plot(Time, NO2/20, '-g','linewidth',2);
plot(Time, HCHO/40, '-k','linewidth',2);
plot(Time, IR340_675/40, '-m','linewidth',2);
plot(Time, I500/6, '-c','linewidth',2);
plot(Time, PC3/40, '-r','linewidth',2);
legend('SO2conc / 35', 'NO2conc / 20', 'HCHOconc / 40',...
'IR_{340/675} / 40', 'I500 / 6', 'PC3 / 40');

set(gcf, 'papersize', [17, 9.6], 'paperposition', [0 0 17 9.6]);
print Newpca3.pdf;

%

plot(Time, SO2/40, '-b','linewidth',2); 
xlabel('Time'); ylabel('');
title('Rescaled standardized SO2/NO2/HCHO, I500, IR340/675 and PC4');
xlim([Time(1), Time(end)]); ylim([-0.4,0.8]);
hold on;
plot(Time, NO2/20, '-g','linewidth',2);
plot(Time, HCHO/35, '-k','linewidth',2);
plot(Time, IR340_675/30, '-m','linewidth',2);
plot(Time, I500/6, '-c','linewidth',2);
plot(Time, PC4/25, '-r','linewidth',2);
legend('SO2conc / 40', 'NO2conc / 20', 'HCHOconc / 35',...
'IR_{340/675} / 30', 'I500 / 6', 'PC4 / 25');

set(gcf, 'papersize', [17, 9.6], 'paperposition', [0 0 17 9.6]);
print Newpca4.pdf;

%

plot(Time, SO2/40, '-b','linewidth',2); 
xlabel('Time'); ylabel('');
title('Rescaled standardized SO2/NO2/HCHO, I500, IR340/675 and PC5');
xlim([Time(1), Time(end)]); ylim([-0.4,0.8]);
hold on;
plot(Time, NO2/20, '-g','linewidth',2);
plot(Time, HCHO/30, '-k','linewidth',2);
plot(Time, IR340_675/20, '-m','linewidth',2);
plot(Time, I500/6, '-c','linewidth',2);
plot(Time, PC5/15, '-r','linewidth',2);
legend('SO2conc / 40', 'NO2conc / 20', 'HCHOconc / 30',...
'IR_{340/675} / 20', 'I500 / 6', 'PC5 / 15');

set(gcf, 'papersize', [17, 9.6], 'paperposition', [0 0 17 9.6]);
print Newpca5.pdf;

% Scatter plots
figure;
scatter(SO2,PC1,'.');
xlabel('SO2');ylabel('PC1');title('SO2 vs PC1');

figure;
scatter(NO2,PC1,'.');
xlabel('NO2');ylabel('PC1');title('NO2 vs PC1');

figure;
scatter(HCHO,PC1,'.');
xlabel('HCHO');ylabel('PC1');title('HCHO vs PC1');

figure;
scatter(IR340_675,PC1,'.');
xlabel('IR_{340/675}');ylabel('PC1');title('IR_{340/675} vs PC1');

figure;
scatter(I500,PC1,'.');
xlabel('I500');ylabel('PC1');title('I500 vs PC1');

%
figure;
scatter(SO2,PC2,'.');
xlabel('SO2');ylabel('PC2');title('SO2 vs PC2');

figure;
scatter(NO2,PC2,'.');
xlabel('NO2');ylabel('PC2');title('NO2 vs PC2');

figure;
scatter(HCHO,PC2,'.');
xlabel('HCHO');ylabel('PC2');title('HCHO vs PC2');

figure;
scatter(IR340_675,PC2,'.');
xlabel('IR_{340/675}');ylabel('PC2');title('IR_{340/675} vs PC2');

figure;
scatter(I500,PC2,'.');
xlabel('I500');ylabel('PC2');title('I500 vs PC2');

%
figure;
scatter(SO2,PC3,'.');
xlabel('SO2');ylabel('PC3');title('SO2 vs PC3');

figure;
scatter(NO2,PC3,'.');
xlabel('NO2');ylabel('PC3');title('NO2 vs PC3');

figure;
scatter(HCHO,PC3,'.');
xlabel('HCHO');ylabel('PC3');title('HCHO vs PC3');

figure;
scatter(IR340_675,PC3,'.');
xlabel('IR_{340/675}');ylabel('PC3');title('IR_{340/675} vs PC3');

figure;
scatter(I500,PC3,'.');
xlabel('I500');ylabel('PC3');title('I500 vs PC3');

%
figure;
scatter(SO2,PC4,'.');
xlabel('SO2');ylabel('PC4');title('SO2 vs PC4');

figure;
scatter(NO2,PC4,'.');
xlabel('NO2');ylabel('PC4');title('NO2 vs PC4');

figure;
scatter(HCHO,PC4,'.');
xlabel('HCHO');ylabel('PC4');title('HCHO vs PC4');

figure;
scatter(IR340_675,PC4,'.');
xlabel('IR_{340/675}');ylabel('PC4');title('IR_{340/675} vs PC4');

figure;
scatter(I500,PC4,'.');
xlabel('I500');ylabel('PC4');title('I500 vs PC4');

%
figure;
scatter(SO2,PC5,'.');
xlabel('SO2');ylabel('PC5');title('SO2 vs PC5');

figure;
scatter(NO2,PC5,'.');
xlabel('NO2');ylabel('PC5');title('NO2 vs PC5');

figure;
scatter(HCHO,PC5,'.');
xlabel('HCHO');ylabel('PC5');title('HCHO vs PC5');

figure;
scatter(IR340_675,PC5,'.');
xlabel('IR_{340/675}');ylabel('PC5');title('IR_{340/675} vs PC5');

figure;
scatter(I500,PC5,'.');
xlabel('I500');ylabel('PC5');title('I500 vs PC5');

File: newPCA.R
# New PCA for data sets of IRI and DOAS .
# 2017-11-06, 2nd version by Weihua Wang
# Spikes filtering first, merge all data later and PCA last.

setwd("/home/wien/")

day=8; month=5; tempdir="085555"; numRows=14716 

source("./findBegEndtime.R")
BegEnd <- findBegEndtime(day, month, tempdir, numRows)
START_MEASUREMENT <- BegEnd[1]
END_MEASUREMENT <- BegEnd[2]

source("./calTOTime.R")
totime <- calTOTime(START_MEASUREMENT, END_MEASUREMENT)
totime

# Read IR(340/675, 440/500, etc) and I500 from a saved dat file calculated by Octave.
IRI <- read.table('IR0508.dat', header = F, stringsAsFactors = FALSE,
                  col.names = c("IR_310T340", "IR_340T440", "IR_340T500", "IR_340T675", "IR_340T870", "IR_440T500",
                                "IR_440T675", "IR_440T870", "IR_500T675", "IR_500T870", "IR_675T870", "I_500"))
#class(IRI$IR_340T675)

timeratio = totime / dim(IRI)[1]
timeratio

# Set time column for IRI
Time <- timeratio*c(1:dim(IRI)[1])  # make up time vector from 0s to 29800s (timeratio*dim(IRI)[1]).
Time <- data.frame(Time)  # add time as a new column into IRI

# Spikes filtering
source("./HpFilterIRI.R")
kk=3
IRI_t <- cbind(IRI,Time)       # _t means time added
IRI.F <- HpFilterIRI(kk,IRI_t) # .F means filtered

# Set another time variable with format of "XX:XX:XX"
TSTD <- read.csv("Time0508.csv", sep = ",", col.names = c("Tstart","Tend"),
                 stringsAsFactors = FALSE) # 14716 X 2
#class(TSTD$Tstart)

IRI.F.T <- cbind(IRI.F, TSTD)  # .T means add another time variable (for merge DOAS data with IR data purpose).

# read csv files with format of e.g. 'DOASeval_YYMMDD_xxxxxx.csv'
fst=52; fed=59

source("./reaDOASfiles.R")
dataset11 <- reaDOASfiles(fst,fed)  
# HCHO at 17th column, NO2a at 20th column, SO2 at 23th column, O3 at 24th column
doasdat <- dataset11[1:dim(dataset11)[1], c(2,17,20,23)]
apply(doasdat, 2, var)

# make a time series column for DOAS data
START_0 = "8:56:38"; END_0 = "17:12:23"
time0 <- as.numeric( toString(difftime(strptime(END_0, format="%H:%M:%OS"), strptime(START_0, format="%H:%M:%OS"), units = "secs") ))
tr0 = time0 / dim(doasdat)[1]
doasdat$Time <- tr0*(1:dim(doasdat)[1])   # add time as a new column into doasdat 

# filter the NO2,HCHO. But SO2 unfiltered for keeping those peak values.
setwd("/home/wien/")
source("./SpikeFilter.R")
width = 1

doasdat.s <- doasdat[3045:dim(doasdat)[1], ] # .s menas sub-set of doasdat

plot(doasdat.s$Time, doasdat.s$SO2conc, type = "l", col='blue', xlab='time', ylab='concentration', main='SO2 concentration')
doasdat.s$SO2conc[doasdat.s$SO2conc < 0.01 ] <- 0.01 # throw values less than 0.01.
plot(doasdat.s$Time, doasdat.s$SO2conc, type = "l", col='blue', xlab='time', ylab='concentration', main='SO2 concentration')

doasdat.s$NO2aconc[doasdat.s$NO2aconc < 0.01 ] <- 0.01
plot(doasdat.s$Time, doasdat.s$NO2aconc, type = "l", col='green', xlab="time", ylab="", main='NO2 concentration')
NO2aconc <- SpikeFilter(doasdat.s$NO2aconc,width)
doasdat.s$NO2aconc <- NO2aconc[[1]]
lines(doasdat.s$Time, doasdat.s$NO2aconc, type = "l", col='red', xlab='time', ylab='concentration', main='Filtered NO2 concentration')

doasdat.s$HCHOaconc[doasdat.s$HCHOaconc < 0.01 ] <- 0.01
plot(doasdat.s$Time, doasdat.s$HCHOaconc, type = "l", col='black', xlab="time", ylab="", main='HCHO concentration')
HCHOaconc <- SpikeFilter(doasdat.s$HCHOaconc,width)
doasdat.s$HCHOaconc <- HCHOaconc[[1]]
lines(doasdat.s$Time, doasdat.s$HCHOaconc, type = "l", col='red', xlab='time', ylab='concentration', main='Filtered NO2 concentration')

apply(doasdat.s, 2, var)  # Now SO2 has the largest variance

#class(doasdat.s$Time); class(doasdat.s$Daytime)
doasdat.s$Daytime <- as.numeric( doasdat.s$Daytime )

#class(IRI.F.T$Tstart)
# Convert a charactor matrix with string format of "xx:xx:xx" to string format of "xxxxx". e.g. "08:52:01" to "85201". 
source("./strVecfromtimeStrvec.R")
IRI.F.T$Tstart <- strVecfromtimeStrvec(IRI.F.T$Tstart)
#class(IRI.F.T$Tstart)  # "character"
#IRI.F.T$Tstart[1]      # "85555"
IRI.F.T$Tend <- strVecfromtimeStrvec(IRI.F.T$Tend)
#IRI.F.T$Tend[dim(IRI)[1]]  # "171235"


IRI.F.T$Tstart.N <- as.numeric(IRI.F.T$Tstart)  # .N means numeric 
IRI.F.T$Tend.N <- as.numeric(IRI.F.T$Tend)
#class(IRI.F.T$Tend.N)

# the merge will throw some data away!
md <- merge.data.frame(doasdat.s, IRI.F.T[,c(4,12,16)], by.x = "Daytime", by.y = "Tstart.N")
apply(md, 2, var)  # check the variance 

# Normalization (because they have different dimensions) before put into PCA, while .n means normalized data set
md.n <- data.frame( scale(md[, -c(1,5),]) ) # exclude Daytime and Time column
apply(md.n, 2, var)  # variance should all be 1

setwd("/home/wien/")
source("./PCA.R")
dim(md.n)[2]

PC <- PCA(md.n[,], "none")

md.n$PC1 <- PC[,1]
md.n$PC2 <- PC[,2]
md.n$PC3 <- PC[,3]
md.n$PC4 <- PC[,4]
md.n$PC5 <- PC[,5]

md.n$Time <- md$Time  # add Time info to normalized md 

# export the result for Octave because adding legend in R is a bit disgusting.
write.csv(md.n,"pcadat1.csv",row.names = FALSE) 
             
# plots only regarding PC1, PC2 and PC3             
plot(md.n$Time, md.n$SO2conc/60, ylim=c(-0.4,0.6), type = "l", col='blue', xlab="Time", ylab="", main='Rescaled standardized SO2/NO2/HCHO, I500, IR340/675 and PC1') 
lines(md.n$Time, md.n$NO2aconc/40, type = "l", col='green', xlab="", ylab="")
lines(md.n$Time, md.n$HCHOaconc/60, type = "l", col='black', xlab="", ylab="")
lines(md.n$Time, md.n$IR_340T675/40, type = "l", col="#CC0000", xlab="", ylab="")
lines(md.n$Time, md.n$I_500/6, type = "l", col='cyan', xlab="", ylab="")
lines(md.n$Time, md.n$PC1/25, type = "l", col="#CC79A7", xlab="", ylab="")
##legend("topright", lwd=1, col=c("blue", "green", 'black', "red"), legend=c("SO2conc/60", "NO2conc/40", "HCHOconc/60", "IR_340T675/40", "I_500/6", "PC2/25"))
legend("topright", lwd=1, col=c("blue", "green", 'black', "red"), legend=c("SO2", "NO2", "HCHO", "IR_340T675", "I_500", "PC2"))

plot(md.n$SO2conc/60, md.n$PC1/40, type="p", col="blue", xlab="SO2", ylab="PC1", main="Scatter plot of SO2 and PC1")
plot(md.n$NO2aconc/40, md.n$PC1/25, type="p", col="blue", xlab="NO2", ylab="PC1", main="Scatter plot of NO2 and PC1")
plot(md.n$HCHOaconc/60, md.n$PC1/25, type="p", col="blue", xlab="HCHO", ylab="PC1", main="Scatter plot of NO2 and PC1")
plot(md.n$IR_340T675/40, md.n$PC1/25, type="p", col="blue", xlab="IR340T675", ylab="PC1", main="Scatter plot of IR340T675 and PC1")
plot(md.n$I_500/6, md.n$PC1/25, type="p", col="blue", xlab="I500", ylab="PC1", main="Scatter plot of I500 and PC1")

#

plot(md.n$Time, md.n$SO2conc/60, ylim=c(-0.4,0.5), type = "l", col='blue', xlab="Time", ylab="", main='Rescaled standardized SO2/NO2/HCHOconc. PC2') 
lines(md.n$Time, md.n$NO2aconc/40, type = "l", col='green', xlab="", ylab="")
lines(md.n$Time, md.n$HCHOaconc/60, type = "l", col='black', xlab="", ylab="")
lines(md.n$Time, md.n$IR_340T675/40, type = "l", col="#CC0000", xlab="", ylab="")
lines(md.n$Time, md.n$I_500/6, type = "l", col='cyan', xlab="", ylab="")
lines(md.n$Time, md.n$PC2/30, type = "l", col="#CC79A7", xlab="", ylab="")
legend("topright", lwd=1, col=c("blue", "green", 'black', "red"), legend=c("SO2conc/60", "NO2conc/40", "HCHOconc/60", "IR_340T675/40", "I_500/6", "PC2/30") )

plot(md.n$SO2conc/60, md.n$PC2/30, type="p", col="blue", xlab="SO2", ylab="PC2", main="Scatter plot of SO2 and PC2")
plot(md.n$NO2aconc/40, md.n$PC2/30, type="p", col="blue", xlab="NO2", ylab="PC2", main="Scatter plot of NO2 and PC2")
plot(md.n$HCHOaconc/60, md.n$PC2/30, type="p", col="blue", xlab="HCHO", ylab="PC2", main="Scatter plot of NO2 and PC2")
plot(md.n$IR_340T675/40, md.n$PC2/30, type="p", col="blue", xlab="IR340T675", ylab="PC2", main="Scatter plot of IR340T675 and PC2")
plot(md.n$I_500/6, md.n$PC2/30, type="p", col="blue", xlab="I500", ylab="PC2", main="Scatter plot of I500 and PC2")

#

plot(md.n$Time, md.n$SO2conc/60, ylim=c(-0.4,0.6), type = "l", col='blue', xlab="Time", ylab="", main='Rescaled standardized SO2/NO2/HCHOconc. PC3') 
lines(md.n$Time, md.n$NO2aconc/40, type = "l", col='green', xlab="", ylab="")
lines(md.n$Time, md.n$HCHOaconc/60, type = "l", col='black', xlab="", ylab="")
lines(md.n$Time, md.n$IR_340T675/40, type = "l", col="#CC0000", xlab="", ylab="")
lines(md.n$Time, md.n$I_500/6, type = "l", col='cyan', xlab="", ylab="")
lines(md.n$Time, md.n$PC3/35, type = "l", col="#CC79A7", xlab="", ylab="")
legend("topright", lwd=1, col=c("blue", "green", 'black', "red"), legend=c("SO2conc/60", "NO2conc/40", "HCHOconc/60", "IR_340T675/40", "I_500/6", "PC3/35") )

plot(md.n$SO2conc/60, md.n$PC3/35, type="p", col="blue", xlab="SO2", ylab="PC3", main="Scatter plot of SO2 and PC3")
plot(md.n$NO2aconc/40, md.n$PC3/35, type="p", col="blue", xlab="NO2", ylab="PC3", main="Scatter plot of NO2 and PC3")
plot(md.n$HCHOaconc/60, md.n$PC3/35, type="p", col="blue", xlab="HCHO", ylab="PC3", main="Scatter plot of NO2 and PC3")
plot(md.n$IR_340T675/40, md.n$PC3/35, type="p", col="blue", xlab="IR340T675", ylab="PC3", main="Scatter plot of IR340T675 and PC3")
plot(md.n$I_500/6, md.n$PC3/35, type="p", col="blue", xlab="I500", ylab="PC3", main="Scatter plot of I500 and PC3")

